ignite/contrib/metrics/classification_report.py

```diff
@@ -0,0 +1,71 @@
+from typing import Callable, List, Optional, Union
+
+import torch
+
+from ignite.metrics import EpochMetric
+
+
+class ClassificationReport(EpochMetric):
+    """Build a text report showing the main classification metrics. applying
+    `sklearn.metrics.classification_report https://scikit-learn.org/stable/modules/generated/
+    sklearn.metrics.classification_report.html`_ .
+
+    Args:
+        output_transform: a callable that is used to transform the
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+        labels: optional list of label indices to include in the report
+        check_compute_fn: Default False. If True, `classification-report
+            <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification-report.html>`_
```

Comment:
Bad link here

Reply:
@vfdev-5 I thought of reusing the existing metrics we have, but the sklearn classification report is more mature and also add a lot of other dimensions including support/averages and some other functionality. for me it made sense to include it like the other contrib metrics we already have.

Reply:
@Moh-Yakoub Thank you for your work. However, I'm not sure that using `sklearn` and `EpochMetric` is the best option. It implies much communication and having all predictions locally. Think of `ImageNet` usecase which could benefits from this report. 

Use our metrics (device aware, distributed, based on PyTorch tensors) makes really more sense for me. The main point is report in a smart way the results from `Precision`, `Recall` and `F1Score` (per class and averaged). Using `lambdaMetric` as @vfdev-5 mentioned would be straightforward.

What do you think ?

Reply:
I will prepare an implementation with `lambdaMetric` and test how it would look like.

Reply:
@sdesrozis  @vfdev-5  I was thinking about that implementation using `MetricsLambda`

At the moment the issues I am having are the following:

1. We don't support precision/recall per label. The main functionality of the sklean method is to generate those metrics per label. that means the only way to go around this in our library is to probably use an `output_transform`.

2. MetricsLambda doesn't automatically update recursively the underlying metrics (https://pytorch.org/ignite/master/generated/ignite.metrics.metrics_lambda.MetricsLambda.html) that means we have to update them by extending the update method. That's not a big issue though and straightforward to implement.

Let me know what you think of this

Reply:
@Moh-Yakoub I agree that this PR implies some improvements.

First, I suggest to write a small piece of code doing the report using our metrics and maybe some specific parts as you mentioned. We will see in a second stage what would be designed in a metric and how we would do it.

What do you think ?

Reply:
> We don't support precision/recall per label. 

This is incorrect, please take a look here: https://pytorch.org/ignite/master/generated/ignite.metrics.precision.Precision.html attribute `average`.

Reply:
@vfdev-5 I think there's a misunderstanding due to the confusion between classes and labels. 

What I meant is this:

Using Sklearn.
```
y_true = [1, 1, 0, 1, 1]
y_pred = [0, 1, 1, 0, 0]
classification_report(y_true, y_pred, output_dict= True)
{'0': {'precision': 0.0, 'recall': 0.0, 'f1-score': 0.0, 'support': 1}, '1': {'precision': 0.5, 'recall': 0.25, 'f1-score': 0.3333333333333333, 'support': 4}, 'accuracy': 0.2, 'macro avg': {'precision': 0.25, 'recall': 0.125, 'f1-score': 0.16666666666666666, 'support': 5}, 'weighted avg': {'precision': 0.4, 'recall': 0.2, 'f1-score': 0.26666666666666666, 'support': 5}}
```

Now, if I use a metrics lambda using `Precision` and `Recall` with `false` `average`s what I get is this

```
{'precision': tensor([0.2500], dtype=torch.float64), 'recall': tensor([0.5000], dtype=torch.float64)}
```

So what happens is that yes, they returned a tensor but in reality it didn't support it per label - unless I am mis-configuring something here.

Reply:
@Moh-Yakoub if you'd like to treat `y_pred` as multiclass and not binary it should have `(N, 2)` shape:
```python
import torch

from ignite.metrics import Precision, Recall


pr = Precision(average=False)
re = Recall(average=False)
f1 = pr * re * 2 / (pr + re + 1e-20)

y_true = torch.tensor([1, 1, 0, 1, 1])
y_pred = torch.tensor([
    [1.0, 0.0], 
    [0.0, 1.0],
    [0.0, 1.0],
    [1.0, 0.0],
    [1.0, 0.0]
])

pr.update((y_pred, y_true))
re.update((y_pred, y_true))

print("Precision per class:", pr.compute())
print("Recall per class:", re.compute())
print("F1 per class:", f1.compute())
```
>
```
Precision per class: tensor([0.0000, 0.5000], dtype=torch.float64)
Recall per class: tensor([0.0000, 0.2500], dtype=torch.float64)
F1 per class: tensor([0.0000, 0.3333], dtype=torch.float64)
```

Reply:
@vfdev-5  @sdesrozis  I've updated the implementation to use our functions + metrics lambda. I'd appreciate if you can take a look at it. once you approve the approach, I will add documentation and fully implement all tests.

Reply:
@Moh-Yakoub Your implementation looks good !